dense AD: introduce convenience functions

andlaus opened the pull request:
this PR was inspired by one of @atgeirr's recent comments. it allows
to get rid if the `MathToolbox<Eval>` detour for the dense AD code in
most cases: e.g. instead of writing

```c++
template <class Eval>
Eval f(const Eval& val)
{ return Opm::MathToolbox<Eval>::sqrt(val); }
```

one can simply write

```c++
template <class Eval>
Eval f(const Eval& val)
{ return Opm::sqrt(val); }
```

and it will work transparently with DenseAD `Evaluation` objects and
primitive floating point values.

one complication of this is that the type of the `Evaluation` object
does not need to be explicitly defined. for functions which take more
than one argument (like e.g. `pow()`, `min()` and `max()`), it is thus
assumed that the type of the result is the same as the type of the
first argument.

another drawback is that when both, the contents of the `Opm::` and
the `std::` namespaces are implicitly available, it is not clear to me
what's used for primitive floating point values. (it seems to compile
but IMO it could lead to surprising behaviour. thus, please only merge
if you consider the benefits of these convenience functions to be
greater than their drawbacks.)

atgeirr commented:
I like this! Also, I think that the answer to your issue about Opm:: and std:: is that a matching non-template overload would be chosen, but I am not 100% sure.

The test seems a bit weak though, it does not check the returned values and while I guess instantiation is performed the if (false) part would probably cause the whole part to go away with optimizations turned on.

jokva commented:
It will always select the most specific overload. I'll have a look what it'll detect in the case you look for.

jokva commented:
> one complication of this is that the type of the Evaluation object
does not need to be explicitly defined. for functions which take more
than one argument (like e.g. pow(), min() and max()), it is thus
assumed that the type of the result is the same as the type of the
first argument.

Can't you use decltype here?

`auto pow( args... ) -> decltype( pow( args... ) ) { /* */ }`

andlaus commented:
> I like this

press green, then :)

> The test seems a bit weak though, it does not check the returned values and while I guess instantiation is performed the if (false) part would probably cause the whole part to go away with optimizations turned on.

The idea of the extension to the test is to make sure the convenience functions compile, and that the code above that checks that they are  behaving correctly (in the sense that they are consistent with the `std::` functions)

of course this does not ensure that the convenience functions themselves do not contain stupid typos.

> It will always select the most specific overload. I'll have a look what it'll detect in the case you think of.

ah, you're right: since the 'std::' functions are not templates and overloads get preceedence over templates it will pick them if possible...

> Can't you use decltype here?

Good idea, I'll try. (but only tomorrow.)

pgdr commented:
The Travis build is failing here, but I believe that is caused by _master_ is actually failing.

My Redhat7-Jenkins in Bergen failed this night with the following message:

```
For keyword WGPR:PRODU14 at reportstep 31
(days, reference value) and (days, check value) = (310, 985.583) and (310, 985.61)
The absolute deviation is 0.0273438. The tolerance limit is 0.02
The relative deviation is 2.7743e-05. The tolerance limit is 1e-05
Program threw an exception: [/home/travis/build/OPM/opm-output/opm/test_util/summaryRegressionTest.cpp:79] Deviation exceed the limit.
90% tests passed, 2 tests failed out of 21
Total Test time (real) =  89.41 sec
The following tests FAILED:
	 20 - check_RESTART_SPE1CASE2 (Failed)
	 21 - compareECLFiles_SPE9_CP (Failed)
Errors while running CTest
```

I have issued a build of opm-simulators master on Travis, so we'll see if it is actually master failing.

Summarizing, Travis fails on this PR, but the same tests fail on my Jenkins on master.

andlaus commented:
@jokva 

I just checked: decltype does not help because the return type must be passed as a template argument to the `MathToolbox` class. What probably will work (but is too complicated IMO) is to take advantage of the SFINAE rule: if one type can be decayed into the other, it likely that this one is the one which is expected by the user. (that probably go wrong if both can be decayed into each other: for e.g., `int` and `double`, `Opm::pow(2, 1.23)` might produce an integer.)

@pgdr 

it seem that opm-output is to blame (if fails at `summaryRegressionTest.cpp:79`)? I can't see a direct link to opm-material.

jokva commented:
@andlaus `Evaluation1` is the return type from `Mathtoolbox::foo()`? If that's the case then sure, auto-decltype won't actually do anything for you, but then you already know the return type and don't have to guess.

andlaus commented:
actually it is not `MathToolbox::foo()` but `MathToolbox<Evaluation1>::foo()` (mind the template argument ;))

alfbr commented:
> it seems that opm-output is to blame (it fails at summaryRegressionTest.cpp:79)? I can't see a direct link to opm-material.

Indeed, it seems the test fails due to exceeding the limits. Did this just happen now? If so, we may actually have a regression due to a recent merge. If not, we probably need to increase tolerances for the tests.

andlaus commented:
my previous the SFINAE idea is the following:

- add a class `ReturnType<Eval1, Eval2>` which checks if both `Eval` can be decayed into each other; if so it picks the type with "better precision"; else if `Eval1` can be decayed into `Eval2`,  then it picks `Eval1`, else `Eval2`.
- change the signature of the convenience functions to something like

```c++
template <class Eval1, class Eval2>
typename ReturnType<Eval1, Eval2>::type max(const Eval1& a, const Eval2& b)
{ return Opm::MathToolbox<typename ReturnType<Eval1, Eval2>::type>::max(a, b); }
```

I'm pretty sure it would work, but it is a bit too much "internal magic" for my liking. That said I don't mind implementing it if there is general appreciation of that approach.

jokva commented:
> actually it is not MathToolbox::foo() but MathToolbox<Evaluation1>::foo() (mind the template argument ;))

Obviously, I was simplying being brief (and inaccurate).

atgeirr commented:
I think that while the ReturnType stuff could work, it is a bit complicated and we'd have to be really careful (and thorough in testing it) to avoid surprises. Therefore I propose a different solution: use the same type for both parameters. That way we are not adding any surprises, the only "problem" is that the mixed-type versions of pow() etc. get no convenience methods. They will still be available through the MathToolBox template as before, so no great harm.

andlaus commented:
> Therefore I propose a different solution: use the same type for both parameters. That way we are not adding any surprises, the only "problem" is that the mixed-type versions of pow() etc. get no convenience methods.

that works, but it would loose part of the point of the convenience functions: while

```c++
Opm::MathToolbox<Eval>::pow(x, 2.0);
```

would work as expected, 

```c++
Opm::pow(x, 2.0);
```

would produce a compiler error. Since I guess that this is a quite common use case, I'm unsure if the convenience functions are worth having in that case (in the sense that they provide some convenience instead of adding complexity)?

atgeirr commented:
> Since I guess that this is a quite common use case, I'm unsure if the convenience functions are worth having in that case

Then we might just not give convenience versions for those functions? Then you still get a compile error if you try Opm::pow(x, 2.0), of course... I think though that the error you would get in the other version (cannot convert a `double` to an `Evaluation<double, 1234>`) is easy to understand. Another alternative: provide overloads using `(Scalar, Evaluation<Scalar, ...>)` and `(Evaluation<Scalar, ...>, Scalar)` arguments. Should also be safe, right?

andlaus commented:
> Then we might just not give convenience versions for those functions?

yeah, that's also possible, but IMO it looks inconsistent from the user's POV. (you need to explain why only single argument functions can be called like `Opm::abs(a)` but not very common two-argument ones like `Opm::max(a, b)`).

> Another alternative: provide overloads using (Scalar, Evaluation<Scalar, ...>) and (Evaluation<Scalar, ...>, Scalar) arguments. Should also be safe, right?

it would work, but there is a problem: the type of `Scalar` is not determined a-priori: it could be `float`, `double`, `__float128`, `int`, `unsigned`, or another Evaluation type (in the case of nesting these AD objects)...

this could mean that

```c++
pow(x, 2.0);
```

works, but

```c++
pow(x, 2.0q);
```

would not, even if `x` uses `__float128` as its underlying floating point type.

to sum up: I think there are three acceptable options:

1) using the first argument as the return type (non-comprehensive)
2) using the SFINAE magic (complicated)
3) not having convenience functions (more verbose)

I'm ok with either of them, you decide.

atgeirr commented:
I think you are wrong that pow(x, 2.0q) would not work, what about this? (untested)
```
template <class Evaluation>
Evaluation min(const Evaluation& arg1, const typename Evaluation::ValueType& arg2)
{
    return Opm::MathToolbox<Evaluation>::min(arg1, arg2);
}
// ... and similar for first arg a scalar, and both args evaluations.
```

atgeirr commented on opm/material/common/MathToolbox.hpp:
```diff
@@ -203,6 +203,72 @@ struct MathToolbox
+
+template <class Evaluation1, class Evaluation2>
+Evaluation1 min(const Evaluation1& arg1, const Evaluation2& arg2)
+{ return Opm::MathToolbox<Evaluation1>::max(arg1, arg2); }
```
Clearly wrong to use max() here. I think that gives the point about actually testing the values some weight...?

andlaus replied:
right, that's one of these stupid copy-and-pastos. fixed, thanks.

andlaus commented:
I was curious of how complicated the SFINAE stuff was and just went ahead and implemented it. It turned out to be much more simple than anticipated (a pleasant surprise for once): if type A is assignable to B, then return B, else A. I've pushed an update to this regard, if you don't like it I will remove it again. (also, I figured that having a `decay<T>(x)` convenience function would be nice to have, so I added that as well.)

andlaus commented:
ok. I looked over this branch again, and I'm reasonably happy with it. (except for the `Opm::getValue(x)` name).

@atgeirr If you are of the same opinion, please press green.

atgeirr commented:
Looks good, will merge, but first I'd like to understand the difference between `getValue()`, `scalarValue()` and `decay<Type>()`: I assume the intention is that getValue returns the value as-is, and scalarValue goes recursively all the way to the bottom-most Evaluation for it's value, so they would only be different for nested Evaluation instances. Is that correct? If so, I don't really understand the use case for decay() and what it is supposed to do.

andlaus commented:
> but first I'd like to understand the difference between `getValue()`, `scalarValue()` and `decay<Type>()`

- `getValue(x)`: for an `Evaluation` object `x` that's equivalent to `x.value` for a primitive floating point object it is `x`. note that the dense AD framework supports nesting, so `x.value` can be an `Evaluation` object itself
- `scalarValue(x)`: "iterated" `getValue(x)` until the result is a primitive floating point value
- `decay<Type>(x)`: if `Type` is the same type as `x`, then it returns `x` and if it is `x::ValueType` it returns `x.value`. the intention is to carry along the derivatives of `x` if they are available, if they are not, `x` is treated as a constant. one prominent use of this is upwinding: if the upstream cell is the cell interior to the considered face, `Type` is `Evaluation`, else it is treated as a constant. another use case is performance: only `Evaluation`  objects need to be stored, but without loss of computational efficiency they can treated as constants if the derivatives are not requested. e.g., when calculating stuff like the storage term while writing the visualization results, the intensive quantities from the linearization can be re-used.

> I assume the intention is that getValue returns the value as-is, and scalarValue goes recursively all the way to the bottom-most Evaluation for it's value, so they would only be different for nested Evaluation instances. Is that correct?

yes

atgeirr commented:
Thanks for the clarification. I do not yet see the light when it comes to decay(), it seems to me that the others would be at least as clear in intention and usage in the cases you mention, but we can continue that discussion later. It is not new to this PR in any case!

atgeirr commented:
OK, think I got it. In that function UpEval can be double or some Evaluation, but the source data is Evaluation always, so you need to decay to a double only if UpEval is actually double.

andlaus commented:
yes; exactly! (be aware that it also took me quite some time to realize that something like this would be useful, too :))
